service/src/main/java/com/codeforcommunity/processor/ReservationProcessorImpl.java

```diff
@@ -0,0 +1,300 @@
+package com.codeforcommunity.processor;
+
+import static org.jooq.generated.Tables.*;
+
+import com.codeforcommunity.api.IReservationProcessor;
+import com.codeforcommunity.auth.JWTData;
+import com.codeforcommunity.dto.reservation.*;
+import com.codeforcommunity.enums.PrivilegeLevel;
+import com.codeforcommunity.enums.ReservationAction;
+import com.codeforcommunity.enums.TeamRole;
+import com.codeforcommunity.exceptions.*;
+import java.sql.Timestamp;
+import java.util.Optional;
+import org.jooq.DSLContext;
+import org.jooq.generated.tables.records.*;
+
+public class ReservationProcessorImpl implements IReservationProcessor {
+
+  private final DSLContext db;
+
+  public ReservationProcessorImpl(DSLContext db) {
+    this.db = db;
+  }
+
+  public Optional<ReservationsRecord> lastAction(int blockId) {
+    return Optional.ofNullable(
+        db.selectFrom(RESERVATIONS)
+            .where(RESERVATIONS.BLOCK_ID.eq(blockId))
+            .orderBy(RESERVATIONS.PERFORMED_AT.desc())
+            .limit(1)
+            .fetchOne());
+  }
+
+  private void basicChecks(int blockId, Integer userId, Integer teamId) {
+    if (!db.fetchExists(db.selectFrom(BLOCKS).where(BLOCKS.ID.eq(blockId)))) {
+      throw new ResourceDoesNotExistException(blockId, "block");
+    }
+
+    if (userId != null) {
+      if (!db.fetchExists(db.selectFrom(USERS).where(USERS.ID.eq(userId)))) {
+        throw new UserDoesNotExistException(userId);
+      }
+    }
+
+    if (teamId != null) {
+      if (!db.fetchExists(db.selectFrom(TEAMS).where(TEAMS.ID.eq(teamId)))) {
+        throw new ResourceDoesNotExistException(teamId, "team");
+      }
+    }
+
+    if (userId != null && teamId != null) {
+      if (!db.fetchExists(
+          db.selectFrom(USERS_TEAMS)
+              .where(USERS_TEAMS.USER_ID.eq(userId))
+              .and(USERS_TEAMS.TEAM_ID.eq(teamId))
+              .and(
+                  USERS_TEAMS
+                      .TEAM_ROLE
+                      .eq(TeamRole.MEMBER)
```

Comment:
I would prefer if we changed this to checking if TEAM_ROLE is **not** PENDING. This is because pending is kind of a special case and it's much more likely we might add another valid team role (like admin or super member, ext...) than it would be to add a role that behaves like pending. That means we would have to know to change all these calls as well which is an easy way to get us in trouble.

Reply:
We could also factor this clause into it's own helper function somewhere: `isOnTeam(int userId, int teamId)` so that we only have a single source of truth for this clause (which is always preferable)

Reply:
Just read the doc and realized you also have to include **not** NONE but I still think reversing the phrasing is worth it

Reply:
This change has been made